doc/tutorials/arrays.md

```diff
@@ -0,0 +1,135 @@
+#> Create a new key user/tests/parent/son with null value
+
+kdb set user/tests/parent/daughter
+#> Create a new key user/tests/parent/daughter with null value
+
+kdb set user/tests/parent/daughter/grandchild
+#> Create a new key user/tests/parent/daughter/grandchild with null value
+```
+
+. We can check the hierarchy of the keys using `kdb ls` and retrieve data using `kdb get`:
+
+```sh
+# Elektra sorts keys alphabetically
+kdb ls user/tests/parent
+#> user/tests/parent
+#> user/tests/parent/daughter
+#> user/tests/parent/daughter/grandchild
+#> user/tests/parent/son
+
+kdb get user/tests/parent
+#> value
+kdb get user/tests/parent/daughter
+#>
+```
+
+.
+
+## Array Keys
+
+Since Elektra sorts keys alphabetically we can use the key-value pair structure described above to store sequences of values (aka. arrays).
+
+### Empty Arrays
+
+For an **empty array** (`[]`) we just add the [meta-key](../help/elektra-metadata.md) `array`:
+
+```sh
+# Create an empty array with the name `user/tests/sequence`
+kdb setmeta user/tests/sequence array ''
+```
+
+.
+
+### Array Elements
+
+To create an **array element** we start the basename of a key with the `#` character and add the index of the array element afterwards. For example, the commands below adds three elements to our array `user/tests/sequence`:
+
+```sh
+kdb set user/tests/sequence/#0 'First Element'
+kdb set user/tests/sequence/#1 'Second Element'
+# Arrays do not need to be contiguous
+kdb set user/tests/sequence/#3 'Fourth Element'
+```
+
+. As you can see above arrays can contain “empty fields”: The key `user/tests/sequence/#2` is missing.
+
+For array elements with an index larger than `9` we must add **underscores** (`_`) to the basename, so we do not destroy the alphabetic order of the array. For example, to add a eleventh element to our array we use the following command:
+
+```sh
+kdb set user/tests/sequence/#_10 'Eleventh Element'
+```
+
+. The order of the array sequence is still correct afterwards, as the following command shows:
+
+```sh
+# List all array elements
+kdb ls user/tests/sequence/
+#> user/tests/sequence/#0
+#> user/tests/sequence/#1
+#> user/tests/sequence/#3
+#> user/tests/sequence/#_10
+```
+
+. For larger indices we add **one underscore less, than the number of digits** of the index. For example, to add a element with index `1337` (`4` digits) we use the basename `#___1337`. We can also generate the basename programmatically:
+
+```bash
+ruby -e 'print("#", "_" * (ARGV[0].length - 1), ARGV[0])' 12345
+#> #____12345
+
+ruby -e 'print("#", "_" * (ARGV[0].length - 1), ARGV[0])' 0
+#> #0
+
+ruby -e 'print("#", "_" * (ARGV[0].length - 1), ARGV[0])' 42
+#> #_42
+```
+
+.
+
+### Metadata
+
+To make life easier for users, Elektra’s [storage plugins](plugins.md) should save the [basename of the last key in the array parent](../decisions/array.md). Some plugins do not support this feature. In that case you can add this value manually via `kdb setmeta`:
```

Comment:
I would say it is a bug if they do not do it, not an unsupported feature.

Reply:
In that case our default storage plugin (`dump`) is buggy:

```sh
kdb setmeta user/tests/array array ''
kdb set user/tests/array/#0 first
kdb getmeta user/tests/array array
#>
```

. By the way, since we are already talking about it, I think it would make sense to specify the features a storage plugins supports (e.g. arrays, comments, metadata, …) via the contract of the plugin. We could then add tests for each of these clauses, to check if the plugin supports a feature properly.

Reply:
In Elektra 0.7 we tried to specify the features of storage plugins but this was not very useful. It is much too difficult for applications to find out if a storage plugin is suitable for their needs.

Instead plugins should now specify the features they lack in `infos/needs`. And special plugins should work around these problems. (like you did with directoryvalue)

So for the problem you mentioned, a plugin should check for validity of arrays. This is, however, more important for yajl than dump, where you currently can even corrupt configuration files if you mix arrays with other keys. But yes, also for dump we should reject broken arrays (or fix them in simple scenarios like you described).